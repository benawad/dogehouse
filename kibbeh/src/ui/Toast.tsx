import * as React from "react";
import { SolidPlus } from "../icons";

export type ToastDurations = "default" | "sticky";

export interface ErrorMessageProps {
  message: string;
  button?: React.ReactNode;
  duration?: ToastDurations;
  onClose?: () => void;
}

export const ErrorToast: React.FC<ErrorMessageProps> = ({
  message,
  button,
  duration = "default",
  onClose,
}) => {
  const onCloseRef = React.useRef(onClose);
  onCloseRef.current = onClose;
  React.useEffect(() => {
    if (duration === "sticky") {
      return;
    }

    const timer = setTimeout(() => {
      onCloseRef.current?.();
    }, 7000);

    return () => {
      clearTimeout(timer);
    };
  }, [duration]);

  return (
    <div
      className={`flex rounded-8 p-3 relative w-full items-center justify-center text-button transition-transform duration-300 bg-secondary`}
      data-testid="error-message"
    >
      {onClose ? (
        <div
          className={`flex absolute cursor-pointer`}
          style={{
            top: 5,
            right: 7,
            width: 13,
            height: 13,
          }}
          onClick={onClose}
          data-testid="close-btn"
        >
          <SolidPlus style={{ transform: "rotate(45deg)" }} />
        </div>
      ) : null}
      <div className={`flex space-x-4 items-center`}>
        <div className={`bold`}>{message}</div>
        {button}
      </div>
    </div>
  );
};
