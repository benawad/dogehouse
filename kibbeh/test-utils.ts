import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import "mutationobserver-shim";

const customRender = (ui: any, options = {}) =>
  render(ui, {
    ...options,
  });

export * from "@testing-library/react";

export { customRender as render };
