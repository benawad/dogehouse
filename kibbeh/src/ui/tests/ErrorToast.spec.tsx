import React from "react";

import { render, act } from "../../../test-utils";
import { ErrorToast } from "../Toast";

jest.useFakeTimers();

describe("ErrorMessage", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render correctly", () => {
    const message = "some error message";
    const { getByTestId, getByText } = render(
      <ErrorToast message={message} onClose={() => {}} />
    );
    const component = getByTestId("error-message");
    const child = getByText(message);

    expect(component).toBeVisible();
    expect(child).toBeVisible();
  });

  it("should call onClose after 7000ms", () => {
    const message = "some error message";
    const onClick = jest.fn();
    render(<ErrorToast message={message} onClose={onClick} />);
    act(() => {
      jest.runAllTimers();
    });

    expect(onClick).toHaveBeenCalled();
  });

  it("shouldn't call onClose after 7000ms if duration props is sticky", () => {
    const message = "some error message";
    const onClick = jest.fn();
    render(
      <ErrorToast message={message} onClose={onClick} duration="sticky" />
    );
    act(() => {
      jest.runAllTimers();
    });

    expect(onClick).not.toHaveBeenCalled();
  });

  it("should call onClose the close button is clicked", () => {
    const message = "some error message";
    const onClick = jest.fn();

    const { getByTestId } = render(
      <ErrorToast message={message} onClose={onClick} />
    );

    const closeBtn = getByTestId("close-btn");
    closeBtn.click();
    expect(onClick).toHaveBeenCalled();
  });

  it("should match snapshot", () => {
    const { getByTestId } = render(
      <ErrorToast message="" onClose={() => {}} />
    );
    const component = getByTestId("error-message");

    expect(component).toMatchSnapshot();
  });
});
