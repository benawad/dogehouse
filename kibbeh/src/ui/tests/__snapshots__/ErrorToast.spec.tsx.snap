// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorMessage should match snapshot 1`] = `
<div
  class="flex rounded-8 p-3 relative w-full items-center justify-center text-button transition-transform duration-300 bg-secondary"
  data-testid="error-message"
>
  <div
    class="flex absolute cursor-pointer"
    data-testid="close-btn"
    style="top: 5px; right: 7px; width: 13px; height: 13px;"
  >
    <svg
      fill="currentColor"
      height="16"
      style="transform: rotate(45deg);"
      viewBox="0 0 16 16"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        clip-path="url(#clip12)"
      >
        <path
          clip-rule="evenodd"
          d="M8 0C8.55228 0 9 0.447715 9 1V15C9 15.5523 8.55228 16 8 16C7.44772 16 7 15.5523 7 15V1C7 0.447715 7.44772 0 8 0Z"
          fill="#DEE3EA"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M0.000976562 8C0.000976562 7.44772 0.448692 7 1.00098 7H15.001C15.5533 7 16.001 7.44772 16.001 8C16.001 8.55228 15.5533 9 15.001 9H1.00098C0.448692 9 0.000976562 8.55228 0.000976562 8Z"
          fill="#DEE3EA"
          fill-rule="evenodd"
        />
      </g>
      <defs>
        <clippath
          id="clip12"
        >
          <rect
            fill="white"
            height="16"
            width="16"
          />
        </clippath>
      </defs>
    </svg>
  </div>
  <div
    class="flex space-x-4 items-center"
  >
    <div
      class="bold"
    />
  </div>
</div>
`;
